File: nodejs-vs-therubyracer/benchmark.rb
require 'v8'

class StopWatch
  def initialize
    @start = Time.now
  end

  def elapsed
    Time.now - @start
  end
end

REPEAT = 10_000
sw = StopWatch.new

REPEAT.times do
  context = V8::Context.new

  # Loads underscore.js
  #underscore = File.read "./files/underscore-min.js"
  underscore = File.read "./files/underscore.js"
  context.eval(underscore)

  # Loads JSON
  json = File.read "./files/input.json"
  context.eval('var json='+json)

  # Runs javascript
  js = File.read "./files/sample.js"
  result = context.eval(js)

  raise "Value is wrong!" if result!=1350
end

puts "Done #{REPEAT} times: #{sw.elapsed} sec"

File: nodejs-vs-therubyracer/benchmark.js
var vm = require('vm'),
    fs = require('fs');

var stopWatch = {
  start: function() {
    this.startTime = new Date();
  },

  elapsed: function() {
    var now = new Date();

    return (now.getTime() - this.startTime.getTime()) / 1000.0;
  }
}

var UNDERSCORE = "./files/underscore-min.js";
var MOMENT = "./files/moment.min.js"
var JSON = "./files/input.json";
var JS = "./files/sample.js";

function benchmark(repeat, withIO) {
  var underscore = fs.readFileSync(UNDERSCORE);
  var moment = fs.readFileSync(MOMENT);
  var json = fs.readFileSync(JSON);
  var js = fs.readFileSync(JS);

  stopWatch.start();
  for (var i=0; i<repeat; i++) {
    var context = vm.createContext();
    var result;

    // underscore
    if (withIO) {
      vm.runInContext(fs.readFileSync(UNDERSCORE), context);
      vm.runInContext(fs.readFileSync(MOMENT), context);
      vm.runInContext('json='+fs.readFileSync(JSON), context);
      result = vm.runInContext(fs.readFileSync(JS), context);

    } else {
      vm.runInContext(underscore, context);
      vm.runInContext(moment, context);
      vm.runInContext('json='+json, context);
      result = vm.runInContext(js, context);
    }

    if (result!=1350) throw new Exception("Value is wrong!");
  }

  var elapsed = stopWatch.elapsed();
  var avg = elapsed / repeat;
  console.log("Done %d times: %d sec. avg: %d (IO=%s)", repeat, elapsed, avg, withIO);
}

benchmark(1000, true);
benchmark(1000, false);

